ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs

```diff
@@ -134,74 +120,66 @@ where
         ZeroKnowledgeProof::new(proof)
     }
 
-    pub fn gen_challenge_and_recurse<λ, I, J>(
+    pub fn gen_challenge_and_recurse<J, B>(
         proof_left: &GenericArray<F, TwoNMinusOne<λ>>,
         proof_right: &GenericArray<F, TwoNMinusOne<λ>>,
-        u: I,
-        v: J,
-    ) -> ProofGenerator<F>
+        mut uv_iterator: J,
+    ) -> Self
     where
-        λ: ArrayLength + Add + Sub<U1>,
+        λ: Add + Sub<U1>,
         <λ as Add>::Output: Sub<U1>,
         <<λ as Add>::Output as Sub<U1>>::Output: ArrayLength,
         <λ as Sub<U1>>::Output: ArrayLength,
-        I: IntoIterator<Item = F>,
-        J: IntoIterator<Item = F>,
-        I::IntoIter: ExactSizeIterator,
-        J::IntoIter: ExactSizeIterator,
+        J: Iterator<Item = B>,
+        B: Borrow<(GenericArray<F, λ>, GenericArray<F, λ>)>,
     {
-        let mut u = u.into_iter();
-        let mut v = v.into_iter();
-
-        debug_assert_eq!(u.len() % λ::USIZE, 0); // We should pad with zeroes eventually
-
-        let s = u.len() / λ::USIZE;
-
-        assert!(
-            s > 1,
-            "When the output is this small, you should validate the proof with a more straightforward reveal"
-        );
-
         let r: F = hash_to_field(
             &compute_hash(proof_left),
             &compute_hash(proof_right),
             λ::U128,
         );
-        let mut p = GenericArray::<F, λ>::generate(|_| F::ZERO);
-        let mut q = GenericArray::<F, λ>::generate(|_| F::ZERO);
+        let mut output = Vec::<(GenericArray<F, λ>, GenericArray<F, λ>)>::new();
         let denominator = CanonicalLagrangeDenominator::<F, λ>::new();
         let lagrange_table_r = LagrangeTable::<F, λ, U1>::new(&denominator, &r);
 
-        let pairs = (0..s).map(|_| {
-            for i in 0..λ::USIZE {
-                let x = u.next().unwrap_or(F::ZERO);
-                let y = v.next().unwrap_or(F::ZERO);
-                p[i] = x;
-                q[i] = y;
+        // iter over chunks of size λ
+        // and interpolate at x coordinate r
+        while let Some(polynomial) = uv_iterator.next() {
+            let mut u = GenericArray::<F, λ>::generate(|_| F::ZERO);
+            let mut v = GenericArray::<F, λ>::generate(|_| F::ZERO);
+            u[0] = lagrange_table_r.eval(&polynomial.borrow().0)[0];
+            v[0] = lagrange_table_r.eval(&polynomial.borrow().1)[0];
+            for i in 1..λ::USIZE {
+                if let Some(polynomial) = uv_iterator.next() {
+                    u[i] = lagrange_table_r.eval(&polynomial.borrow().0)[0];
+                    v[i] = lagrange_table_r.eval(&polynomial.borrow().1)[0];
+                }
```

Comment:
Why treat index 0 and the others differently?

Reply:
Also, I think it would be more readable to use de-structuring as opposed to `.0` and `.1` which I find particularly unreadable.

```
let (u_chunk, v_chunk) = polynomial.borrow();
...
u[i] = lagrange_table_r.eval(u_chunk)[0];
v[i] = lagrange_table_r.eval(v_chunk)[0];
```

Is this possible or do we encounter issues with "borrow"?

Reply:
Just changed it and it works!

Why is 0 treated differently: I do a compression here, basically I compress λ many generic arrays of length λ into a single generic array of length λ. That is why I manually iterate by calling next. Whenever I need to start a new array, I generate a 0 array. Since next is `some`, the first element of the new array always has a value that I can place. Now I need to call another λ-1 times next but the iterator might stop, i.e. next() is none. Then I don't set a value and the current index i and all following indexes are remaining 0, I.e. I fill them with 0 such that the total amount of field elements is a multiple of λ and there is no issue with the next recursion.